lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2382 @@
+                pred_action,
+                action,
+                reduction="mean",
+            )
+            return pred_action, loss
+
+
+class VQBeTHead(nn.Module):
+    def __init__(self, config: VQBeTConfig):
+        """
+        VQBeTHead takes output of GPT layers, and pass the feature through bin prediction head (`self.map_to_cbet_preds_bin`), and offset prediction head (`self.map_to_cbet_preds_offset`)
+
+        self.map_to_cbet_preds_bin: outputs probability of each code (for each layer).
+            The input dimension of `self.map_to_cbet_preds_bin` is same with the output of GPT, 
+            and the output dimension of `self.map_to_cbet_preds_bin` is `self.config.vqvae_groups * self.config.vqvae_n_embed`, where 
+                `self.config.vqvae_groups` is number of RVQ layers, and
+                `self.config.vqvae_n_embed` is codebook size of RVQ.
+
+        self.map_to_cbet_preds_offset: output the predicted offsets for all the codes in all the layers. 
+            The input dimension of ` self.map_to_cbet_preds_offset` is same with the output of GPT, 
+            and the output dimension of ` self.map_to_cbet_preds_offset` is `self.config.vqvae_groups * self.config.vqvae_n_embed * config.n_action_pred_chunk * config.output_shapes["action"][0]`, where
+                `self.config.vqvae_groups` is number of RVQ layers,
+                `self.config.vqvae_n_embed` is codebook size of RVQ,
+                `config.n_action_pred_chunk is action chunk size of each token, and
+                `config.output_shapes["action"][0]` is the dimension of action 
+        """
+
+        super().__init__()
+        self.config = config
+
+
+
+        self.map_to_cbet_preds_bin = MLP(
+            in_channels=config.gpt_output_dim,
+            hidden_channels=[self.config.vqvae_groups * self.config.vqvae_n_embed],
+        )
+        self.map_to_cbet_preds_offset = MLP(
+            in_channels=config.gpt_output_dim,
+            hidden_channels=[
+                self.config.vqvae_groups * self.config.vqvae_n_embed * config.n_action_pred_chunk * config.output_shapes["action"][0],
+            ],
+        )
+        # init vqvae
+        self.vqvae_model = VqVae(config)
+        # loss
+        self._criterion = FocalLoss(gamma=2.0)
+
+    def discretize(self, discretize_step, actions):
+        if next(self.vqvae_model.encoder.parameters()).device != get_device_from_parameters(self):
+            self.vqvae_model.encoder.to(get_device_from_parameters(self))
+            self.vqvae_model.vq_layer.to(get_device_from_parameters(self))
+            self.vqvae_model.decoder.to(get_device_from_parameters(self))
+            self.vqvae_model.device = get_device_from_parameters(self)
+
+        loss, n_different_codes, n_different_combinations = pretrain_vqvae(self.vqvae_model, discretize_step, actions)
+        # if we updated RVQ more than `discretize_step` steps,
+        if self.vqvae_model.discretized:
```

Comment:
Any chance we can move this logic to the top of the stack in `VQBeTPolicy`? There is already a similar check there, and I think it would be clearer to have all this logic in one place.

By the way, the call to `eval` will be soon undermined by the training loop which will call `train` on the top level module. Is that okay? If so, should we remove it here?

Reply:
This part is just for setting vqvae_model as `eval()`.

I moved this part inside `def pretrain_vqvae(vqvae_model, discretize_step, actions):`
Would it make sense?

Reply:
Thanks for that! I'll double check when I get to it :)